test/smtp.test.lua

```diff
@@ -146,6 +166,52 @@ test:test("smtp.client", function(test)
                   m.text,
                   "Subject: =%?utf%-8%?b%?YWJjZGVmZ2hpamvRj2xtbm9wcXJzdHV2d3h5eg==%?=", ""))
     test:is(subj, 1, 'subject codes >127')
-end)
 
+    r = client:request(addr, '3xx@tarantool.org',
+                       'receiver@tarantool.org',
+                       'mail.body')
+    test:is(r.reason, 'MAIL failed: 354', 'Errors 3xx')
+    test:is(r.status, 354, 'Unexpected code')
+
+    r = client:request(addr, '4xx@tarantool.org',
+                       'receiver@tarantool.org',
+                       'mail.body')
+    test:is(r.reason, 'MAIL failed: 421', 'service unavailable')
+    test:is(r.status, 421, 'Unexpected code')
+
+    r = client:request(addr, '5xx@tarantool.org',
+                       'receiver@tarantool.org',
+                       'mail.body')
+    test:is(r.reason, 'MAIL failed: 510', 'unexisting recipient')
+    test:is(r.status, 510, 'Unexpected code')
+
+    r = client:request(addr, 'breakconnect@tarantool.org',
+                       'receiver@tarantool.org',
+                       'mail.body')
+    test:is(r.reason, 'response reading failed', 'unexpected response code')
+
+    r = client:request(addr, 'sender@tarantool.org',
+                       '3xx@tarantool.org',
+                       'mail.body')
+    test:is(r.reason, 'RCPT failed: 354', 'Errors 3xx')
+    test:is(r.status, 354, 'Unexpected code')
+
+    r = client:request(addr, 'sender@tarantool.org',
+                       '4xx@tarantool.org',
+                       'mail.body')
+    test:is(r.reason, 'RCPT failed: 421', 'service unavailable')
+    test:is(r.status, 421, 'Unexpected code')
+
+    r = client:request(addr, 'sender@tarantool.org',
+                       '5xx@tarantool.org',
+                       'mail.body')
+    test:is(r.reason, 'RCPT failed: 510', 'unexisting recipient')
+    test:is(r.status, 510, 'Unexpected code')
+
+    r = client:request(addr, 'sender@tarantool.org',
+                       'breakconnect@tarantool.org',
+                       'mail.body')
+    test:is(r.reason, 'response reading failed', 'unexpected response code')
```

Comment:
What is the idea behind duplicating those test cases with error injection using a sender address and a receiver one?

Reply:
The idea is that there are two phases in SMTP connection, when we can get some SMTP error from server. If server doesn't want to or can't work with sender and if it doesn't want to or can't work with recipient. We should make sure that both cases can be handled.

Reply:
Okay, thanks for the explanation.